### PR TITLE
docs(x/data): update specification

### PR DESCRIPTION
**Pull request by ryanchristo**
## Description

Closes: #571

This pull request updates the data module specification to reflect the latest version.

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct `docs:` prefix in the PR title
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the [documentation writing guidelines](https://github.com/cosmos/cosmos-sdk/blob/master/docs/DOC_WRITING_GUIDELINES.md)
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct `docs:` prefix in the PR title
- [ ] confirmed all author checklist items have been addressed 
- [ ] confirmed that this PR only changes documentation
- [ ] reviewed content for consistency
- [ ] reviewed content for thoroughness
- [ ] reviewed content for spelling and grammar
- [ ] tested instructions (if applicable)

### CODE REVIEW

**Review by robert-zaremba — COMMENTED** [review]
Similar to: https://github.com/regen-network/regen-ledger/pull/633#pullrequestreview-863414089

Let's simplify the acceptance tests

**Comment by ryanchristo** [comment]
> Similar to: #633 (review)
>
> Let's simplify the acceptance tests

I don't think changing the language is necessary. I think we should follow the standard format for acceptance tests: https://openclassrooms.com/en/courses/4544611-write-agile-documentation-user-stories-acceptance-tests/4810081-write-acceptance-tests

**Comment by clevinson** [comment]
> I don't think changing the language is necessary. I think we should follow the standard format for acceptance tests: https://openclassrooms.com/en/courses/4544611-write-agile-documentation-user-stories-acceptance-tests/4810081-write-acceptance-tests

I'm actually not sure from reading this that the "if..then" is part of the standard format. When I read this it makes me think that the "if...then" statements they provide are only to help one generate a GIVEN...WHEN...THEN style acceptance test. I think these are all fine to leave in in the current UATs. In the future I do agree that they can be simpler (5-10 word) summaries of what the condition is. But I don't think this is really anything worth bike shedding over.

**Review by clevinson — APPROVED** [review]
I noticed that the pages weren't actually visible in the table of contents, or in the Data Module overview page. So I added a few commits to fix that. I think its good for these to be visible even though they're not available on main net (and just added a note abt only being available in experimental builds in the data module overview page)

**Comment by ryanchristo** [comment]
Thanks @clevinson. Opened #711 to further discuss and track updates to format.

File: x/data/spec/02_state.md [hunk 1]
```diff
@@ -1 +1,3 @@
 # State
+
+The `data` module uses a lookup table as an abstraction over Cosmos SDK's `KVStore` for storing content hashes that adhere to a canonical regen IRI scheme. The lookup table creates a short, unique binary identifier for the longer content hash using an efficient, non-cryptographic hash function.
```
Comment (technicallyty) [inline x/data/spec/02_state.md:3]:
```suggestion
The `data` module uses a lookup table for storing content hashes that adhere to a canonical regen IRI scheme. The lookup table creates a short, unique binary identifier for the longer content hash using an efficient, non-cryptographic hash function.
```

i feel like the specifics of the storage aren't too important, especially as we migrate towards the ORM, just the general structure of how the data is stored should be good enough imo

File: x/data/spec/03_messages.md [hunk 2]
```diff
@@ -1 +1,30 @@
...
+- an on-chain identity may represent an organization and through delegation individual members may sign on behalf of the group
+- the blockchain transaction envelope provides built-in replay protection and timestamping
+
+`SignData` implicitly calls `AnchorData` if the data was not already anchored. `SignData` can be called multiple times for the same content hash with different signers and those signers will be appended to the list of signers.
+
++++ https://github.com/regen-network/regen-ledger/blob/55d39dc7d9768e5a0bd18c48ca78cb4a02145e81/proto/regen/data/v1alpha2/tx.proto#L64-L77
+
+### Validation:
+
+- `signers` must ba valid addresses, and their signatures must be present in the transaction
```
Comment (technicallyty) [inline x/data/spec/03_messages.md:29]:
```suggestion
- `signers` must be a valid addresses, and their signatures must be present in the transaction
```

File: x/data/spec/05_client.md [hunk 3]
```diff
@@ -1,42 +1,345 @@
...
+
+The `tx` commands allow users to interact with the `data` module.
+
+```bash
+regen tx data --help
+```
+
+#### anchor
+
+The `anchor` command allows users to anchor a data to the blockchain based on its content hash (i.e. IRI).
```
Comment (technicallyty) [inline x/data/spec/05_client.md:108]:
```suggestion
The `anchor` command allows users to anchor data to the blockchain based on its content hash (i.e. IRI).
```

File: x/data/spec/06_tests.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+- GIVEN - the content hash is of graph type and the canonicalization algorithm is not supported
+- WHEN - user tries to anchor data
+- THEN - transaction fails, data is not anchored
+
+If a user tries to anchor data and the content hash is of graph type and the canonicalization algorithm is not supported, then the transaction fails and the data is not anchored.
+
+- GIVEN - the content hash is of graph type and the canonicalization algorithm is not supported
+- WHEN - user tries to anchor data
+- THEN - transaction fails, data is not anchored
```
Comment (technicallyty) [inline x/data/spec/06_tests.md:41]:
```suggestion
If a user tries to anchor data and the content hash is of graph type and the canonicalization algorithm is not supported, then the transaction fails and the data is not anchored.

- GIVEN - the content hash is of graph type and the canonicalization algorithm is not supported
- WHEN - user tries to anchor data
- THEN - transaction fails, data is not anchored
```

looks like there was a duplicate

File: x/data/spec/06_tests.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+If a user tries to anchor data and the content hash is of raw type and the digest length is not 256 bits, then the transaction fails and the data is not anchored.
+
+- GIVEN - the content hash is of raw type and the digest length is not 256 bits
+- WHEN - user tries to anchor data
+- THEN - transaction fails, data is not anchored
+
+### Sign Data
```
Comment (technicallyty) [inline x/data/spec/06_tests.md:67]: should we also include a test where a hash is passed but is not found on the chain?

Comment (ryanchristo) [reply x/data/spec/06_tests.md:61]: added in https://github.com/regen-network/regen-ledger/pull/673/commits/2998bb1e72c25eb9206933924b5e4f35354ba53c

**Review by technicallyty — APPROVED** [review]
👍🏻

**Pull request merged**